lib/SciMLCompat/src/enzyme.jl

```diff
@@ -0,0 +1,278 @@
+""" 
+This module wraps some of the functions of the IncompressibleNavierStokes module
+in order to make them compatible with Enzyme AD.
+
+There are also some tests to check if the redefined functions are faster than the original ones and if they are differentiable.
+"""
+
+using IncompressibleNavierStokes 
+using SparseArrays
+using LinearAlgebra
+using KernelAbstractions
+using Enzyme
+# runtimeActivity is required to tell Enzyme that all temporary variables should be active
+Enzyme.API.runtimeActivity!(true)
+
+struct Offset{D} end
+@inline (::Offset{D})(α) where {D} = CartesianIndex(ntuple(β -> β == α ? 1 : 0, D))
+
+"""
+    _get_enz_bc_u!(_backend, setup)
+
+This function is used to precompile the function that applies the boundary conditions on u.
+It wraps everything such that the only thing that has to be passed from outside is the u array.
+This makes it easy for Enzyme to differentiate the function.
+"""
+function _get_enz_bc_u!(_backend, setup)
+    (; boundary_conditions, grid, workgroupsize) = setup
+    (; dimension, N) = grid
+    D = dimension()
+    e = Offset{D}()
+    for β = 1:D
+        @assert boundary_conditions[β][1] isa PeriodicBC "Only PeriodicBC implemented"
+    end
+
+    @kernel function _bc_a!(u, ::Val{α}, ::Val{β}) where {α,β}
+        I = @index(Global, Cartesian)
+        u[α][I] = u[α][I+(N[β]-2)*e(β)]
+    end
+    @kernel function _bc_b!(u, ::Val{α}, ::Val{β}) where {α,β}
+        I = @index(Global, Cartesian)
+        u[α][I+(N[β]-1)*e(β)] = u[α][I+e(β)]
+    end
+
+    function bc_u!(u, β; isright, )
+        ndrange = ntuple(γ -> γ == β ? 1 : N[γ], D)
+        if isright
+            for α = 1:D
+                _bc_b!(_backend, workgroupsize)(u, Val(α), Val(β); ndrange)
+            end
+        else
+            for α = 1:D
+                _bc_a!(_backend, workgroupsize)(u, Val(α), Val(β); ndrange)
+            end
+        end
+        
+    end
+
+    function bc(u)
```

Comment:
function cannot be proven to be readonly (according to Enzyme). probably because of the call to bc_u! (which does edit u)